keps/sig-windows/2258-node-service-log-viewer/README.md

```diff
@@ -0,0 +1,389 @@
+the existing filters supported by journalctl. The `kubectl logs` will have
+command line options for specifying these filters when interacting with node
+objects.
+
+On the Windows side viewing of logs from services that use `C:\var\log` will
+be supported by the existing endpoint. For Windows services that log to the
+the System and Application logs, Windows Event Logs and Event Tracing (ETW),
+we can leverage the [Get-WinEvent cmdlet](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.diagnostics/get-winevent?view=powershell-7.1)
+that supports getting logs from all these sources. The cmdlet has filtering
+options that can be leveraged to filter the logs in the same manner we do
+with the journal logs.
+
+Please note that filtering will not be available for logs in `/var/log/` or
+`C:\var\log\`.
+
+On clusters where the same service using different logging mechanisms,
+example: kubelet logs to `/var/log` on some nodes and uses the journal on other
+nodes, the cluster administrator will have to use `kubectl logs` with different
+set of options to read both types of logs.
+
+The feature now enables the cluster administrator to interrogate all services.
+This could be prevented by having a whitelist of allowed services. But this
+comes with severe disadvantages as there could be nodes (especially with
+Windows) that have other services to support networking and monitoring.
+These services are variable and will depend on how the nodes have been
+configured. Here are some examples:
+- [hybrid-overlay-node](https://github.com/ovn-org/ovn-kubernetes/tree/master/go-controller/hybrid-overlay)
+- [windows-exporter](https://github.com/prometheus-community/windows_exporter).
+
+
+The `/var/log/` endpoint is enabled using the `enableSystemLogHandler` kubelet
+configuration options. To gain access to this new feature this option needs to
+be enabled. In addition when introducing this feature it will be hidden behind a
+`NodeLogs` feature gate in the kubelet that needs to be explicitly enabled. So
+you need to enable both options to get access to this new feature and disabling
+`enableSystemLogHandler` will disable the new feature irrespective of the
+`NodeLogs` feature gate.
+
+A reference implementation of this feature without the feature gate is
+available [here](https://github.com/kubernetes/kubernetes/pull/96120).
+
+#### kubectl
+
+`kubectl` has an existing `logs` command that is used to view the logs for a
+container in a pod or a specified resource. The sub-command looks at resource
+types, so can be extended to work with node objects to view the logs of services
+on the nodes. Given that the `logs` command depends on RBAC policies for access
+to appropriate resource type and associated endpoints, it will allow us to
+restrict node logs access to only cluster administrators as long as the cluster
+is setup in that manner.
+
+The default mode for `logs` sub-command for node objects would be to query
```

Comment:
Note that the oc command has a fairly sophisticated alignment mechanism when viewing node logs across multiple nodes that is similar to but more powerful than the current log slurping across pods behavior in kubectl logs.  I would note that the implementation of log unification by timestamp is worth preserving and would benefit pod log viewing as well.  We would want to get someone familiar with the pod logs implementation to review and note where they see potential issues:

https://github.com/openshift/oc/blob/master/pkg/cli/admin/node/logs.go#L127
https://github.com/openshift/oc/blob/master/pkg/cli/admin/node/logs.go#L309
https://github.com/openshift/oc/blob/master/pkg/cli/admin/node/logs.go#L429

This makes viewing logs "since" a time across multiple nodes much more palatable, since you can see events aligned by time quickly.

Reply:
I too think we should keep this focused if possible - if more sophistication can be added to the alignment mechanism later, I'd prefer that.